File: src/main/java/net/volcano/jdacommands/model/command/CommandCompiler.java
package net.volcano.jdacommands.model.command;

import net.dv8tion.jda.api.requests.RestAction;
import net.volcano.jdacommands.config.CategoryConfig;
import net.volcano.jdacommands.exceptions.command.CommandCompileException;
import net.volcano.jdacommands.model.ParameterData;
import net.volcano.jdacommands.model.command.annotations.BotOwnerCanAlwaysExecute;
import net.volcano.jdacommands.model.command.annotations.CommandController;
import net.volcano.jdacommands.model.command.annotations.CommandMethod;
import net.volcano.jdacommands.model.command.annotations.Help;
import net.volcano.jdacommands.model.command.arguments.ArgumentList;
import net.volcano.jdacommands.model.command.arguments.CommandArgument;
import net.volcano.jdacommands.model.command.arguments.interfaces.CodecRegistry;
import net.volcano.jdautils.utils.ClassUtil;

import java.lang.reflect.Method;
import java.util.*;

public class CommandCompiler {
	
	private final CodecRegistry registry;
	private final CategoryConfig config;
	
	public CommandCompiler(CodecRegistry registry, CategoryConfig config) {
		this.registry = registry;
		this.config = config;
	}
	
	public Set<Command> compile(Object controller) throws CommandCompileException {
		
		CommandController module = controller.getClass().getAnnotation(CommandController.class);
		
		if (module == null) {
			throw new IllegalArgumentException("Object is not a command!");
		}
		
		var methods = controller.getClass().getDeclaredMethods();
		
		Set<Command> set = new HashSet<>();
		for (Method method : controller.getClass().getDeclaredMethods()) {
			if (method.isAnnotationPresent(CommandMethod.class)) {
				var command = compileCommand(controller, module, method);
				command.arguments.command = command;
				set.add(command);
			}
		}
		return set;
		
	}
	
	public Command compileCommand(Object o, CommandController controller, Method method) throws CommandCompileException {
		
		CommandMethod commandMethod = method.getAnnotation(CommandMethod.class);
		Command.CommandBuilder builder = Command.builder();
		
		if (method.getReturnType() != Void.TYPE && method.getReturnType() != RestAction.class) {
			throw new CommandCompileException(method, "Command method needs to return RestAction or Void!");
		}
		
		List<String> paths = new ArrayList<>();
		for (String controllerPath : controller.path()) {
			for (String commandPath : commandMethod.path()) {
				String path = controllerPath + " " + commandPath;
				path = path.replaceAll(" +", " ").trim();
				if (path.length() == 0) {
					throw new CommandCompileException(method, "Command path cannot be empty");
				}
				paths.add(path);
			}
		}
		builder.paths(paths.toArray(new String[0]));
		
		// Build the arguments for this command
		List<CommandArgument> arguments = new ArrayList<>();
		boolean firstIsEvent = false;
		boolean lastIsArray = false;
		var params = method.getParameters();
		for (int i = 0; i < params.length; i++) {
			
			if (i == 0 && params[i].getType() == CommandEvent.class) {
				firstIsEvent = true;
				continue;
			}
			
			var isArray = params[i].getType().isArray();
			
			Class<?> actualType = params[i].getType();
			if (isArray) {
				actualType = actualType.componentType();
			}
			
			var isEnum = actualType.isEnum();
			var isPrimitive = actualType.isPrimitive();
			//var isGeneric = ((ParameterizedType) ((ParameterizedType) params[i].getType().getGenericSuperclass()).getActualTypeArguments()[0]).getActualTypeArguments()[0] instanceof WildcardType;
			
			if (isArray && i != params.length - 1) {
				throw new CommandCompileException(method, "Cannot have array argument as non last parameter");
			}
			
			Class<?> codecType;
			if (isEnum) {
				codecType = Enum.class;
			} else if (isPrimitive) {
				codecType = ClassUtil.dePrimitivize(params[i].getType());
			} else {
				codecType = actualType;
			}
			
			var codec = registry.getCodec(codecType);
			
			if (codec == null) {
				throw new CommandCompileException(method, "Unsupported type for command argument; " + codecType);
			} else {
				var data = new ParameterData(params[i], actualType, codecType, registry);
				arguments.add(codec.encodeArgument(data));
			}
			
			if (isArray) {
				lastIsArray = true;
			}
			
		}
		
		ArgumentList argumentList = new ArgumentList(arguments, lastIsArray);
		builder.arguments(argumentList);
		
		// Build command help
		
		if (method.isAnnotationPresent(Help.class)) {
			Command.Help.HelpBuilder helpBuilder = Command.Help.builder();
			
			Help help = method.getAnnotation(Help.class);
			
			helpBuilder.usage(help.usage().equals("GENERATE") ? paths.get(0) + " " + argumentList.generateUsage() : help.usage());
			helpBuilder.description(help.description());
			var category = "";
			if (help.category().equals("DEFAULT")) {
				category = controller.category();
			} else {
				category = help.category();
			}
			helpBuilder.category(category);
			if (config.emojis.containsKey(category.toLowerCase())) {
				helpBuilder.emoji(config.emojis.get(category));
			} else {
				throw new CommandCompileException(method, "No emoji for category " + category);
			}
			helpBuilder.examples(help.examples());
			helpBuilder.details(help.details());
			helpBuilder.permissions(Arrays.stream(help.permissions())
					.map(p -> "`" + p.permission() + "` : " + p.description())
					.toArray(String[]::new));
			
			builder.help(helpBuilder.build());
		}
		
		// Build the command method for this command
		CommandFunction.CommandFunctionBuilder methodBuilder = CommandFunction.builder();
		
		methodBuilder.method(method);
		methodBuilder.argumentCount(arguments.size() + (firstIsEvent ? 1 : 0));
		methodBuilder.includeEvent(firstIsEvent);
		methodBuilder.instance(o);
		
		builder.method(methodBuilder.build());
		
		// Build the permissions for this command
		var permissions = controller.permissions();
		if (!permissions.isBlank() && !commandMethod.permissions().isBlank()) {
			permissions += ".";
		}
		permissions += commandMethod.permissions();
		if (permissions.isEmpty()) {
			throw new CommandCompileException(method, "Command permissions are empty, set some and provide them by default instead.");
		} else if (!permissions.startsWith("command.")) {
			permissions = "command." + permissions;
		}
		builder.permission(permissions);
		
		// Add command source
		if (commandMethod.source() == Command.Source.DEFAULT) {
			if (controller.source() == Command.Source.DEFAULT) {
				builder.source(Command.Source.BOTH);
			} else {
				builder.source(controller.source());
			}
		} else {
			builder.source(commandMethod.source());
		}
		
		// Add misc values
		builder.sensitive(commandMethod.sensitive());
		builder.globalPermissions(commandMethod.global());
		
		builder.botOwnerCanAlwaysExecute(method.isAnnotationPresent(BotOwnerCanAlwaysExecute.class));
		
		return builder.build();
		
	}
	
}

File: src/main/java/net/volcano/jdacommands/model/command/CommandCompiler.kt
package net.volcano.jdacommands.model.command

import net.dv8tion.jda.api.requests.RestAction
import net.volcano.jdacommands.config.CategoryConfig
import net.volcano.jdacommands.exceptions.command.CommandCompileException
import net.volcano.jdacommands.model.ClassUtil
import net.volcano.jdacommands.model.ParameterData
import net.volcano.jdacommands.model.command.Command.Companion.CommandBuilder
import net.volcano.jdacommands.model.command.Help.Companion.HelpBuilder
import net.volcano.jdacommands.model.command.annotations.BotOwnerCanAlwaysExecute
import net.volcano.jdacommands.model.command.annotations.CommandController
import net.volcano.jdacommands.model.command.annotations.CommandMethod
import net.volcano.jdacommands.model.command.annotations.Help
import net.volcano.jdacommands.model.command.arguments.ArgumentList
import net.volcano.jdacommands.model.command.arguments.CommandArgument
import net.volcano.jdacommands.model.command.arguments.interfaces.CodecRegistry
import net.volcano.jdautils.utils.isArray
import net.volcano.jdautils.utils.kClass
import kotlin.reflect.KFunction
import kotlin.reflect.full.declaredMemberFunctions
import kotlin.reflect.full.findAnnotation

class CommandCompiler(
	private val registry: CodecRegistry,
	private val config: CategoryConfig
) {

	@Throws(CommandCompileException::class)
	fun compile(controller: Any): Set<Command> {

		val module = controller.javaClass.getAnnotation(CommandController::class.java)
			?: throw IllegalArgumentException("Object is not a command!")

		val set: MutableSet<Command> = HashSet()
		for (method in controller::class.declaredMemberFunctions) {
			if (method.findAnnotation<CommandMethod>() != null) {
				val command = compileCommand(controller, module, method)
				command.arguments.command = command
				set.add(command)
			}
		}

		return set
	}

	@Throws(CommandCompileException::class)
	fun compileCommand(o: Any?, controller: CommandController, function: KFunction<*>): Command {
		val annotation = function.findAnnotation<CommandMethod>()!!
		val builder = CommandBuilder()

		if (function.returnType != Void.TYPE && function.returnType != RestAction::class)
			throw CommandCompileException(function, "Command function needs to return RestAction or Void!");

		val paths = controller.path
			.flatMap { controllerPath ->
				annotation.path.map { commandPath ->
					"$controllerPath $commandPath".replace(Regex("  +"), " ").trim()
				}
			}

		if (paths.any { it.isBlank() }) throw CommandCompileException(function, "Command path cannot be empty.")

		builder.paths = paths.toTypedArray()

		// Build the arguments for this command
		val arguments: MutableList<CommandArgument<*>> = mutableListOf()
		var firstIsEvent = false
		var lastIsArray = false
		val parameters = function.parameters
		for ((index, param) in parameters.withIndex()) {

			if (index == 0 && param.type == CommandEvent::class.java) {
				firstIsEvent = true
				continue
			}

			if (param.type.isArray || param.isVararg) {
				if (index + 1 == function.parameters.size)
					lastIsArray = true
				else
					throw CommandCompileException(function, "Cannot have array/vararg argument as non-last parameter.")
			}

			val codecType = ClassUtil.getCodecClass(param.type.kClass)

			val codec = registry.getCodec(codecType) ?: throw CommandCompileException(
				function,
				"Unsupported type for command argument; $codecType"
			)

			arguments.add(codec.encodeArgument(ParameterData(param, codecType, registry)))

		}
		val argumentList = ArgumentList(arguments, lastIsArray)
		builder.arguments = argumentList
		// Build command help
		function.findAnnotation<Help>()?.let {
			val helpBuilder = HelpBuilder()

			helpBuilder.usage = if (it.usage == "GENERATE") paths[0] + " " + argumentList.generateUsage() else it.usage
			helpBuilder.description = it.description

			val category = if (it.category == "DEFAULT") controller.category else it.category
			helpBuilder.category = category

			helpBuilder.emoji =
				config.emojis[category] ?: throw CommandCompileException(function, "No emoji for category $category")
			helpBuilder.examples = it.examples
			helpBuilder.details = it.details
			helpBuilder.permissions = it.permissions.map { p -> "`${p.permission}` : ${p.description}" }.toTypedArray()

			builder.help = helpBuilder.build()
		}

		// Build the command method for this command
		builder.function = CommandFunction(function, firstIsEvent)

		// Build the permissions for this command
		var permissions = controller.permissions
		if (permissions.isNotBlank() && annotation.permissions.isNotBlank()) permissions += "."
		permissions += annotation.permissions
		if (permissions.isEmpty()) {
			throw CommandCompileException(
				function,
				"Command permissions are empty, set some and provide them by default instead."
			)
		} else if (!permissions.startsWith("command.")) {
			permissions = "command.$permissions"
		}
		builder.permission = permissions

		builder.source = if (annotation.source == Source.DEFAULT)
			if (controller.source == Source.DEFAULT)
				Source.BOTH
			else
				controller.source
		else
			annotation.source

		// Add misc values
		builder.sensitive = annotation.sensitive
		builder.globalPermissions = annotation.global
		builder.botOwnerCanAlwaysExecute = function.findAnnotation<BotOwnerCanAlwaysExecute>() != null

		return builder.build()
	}
}